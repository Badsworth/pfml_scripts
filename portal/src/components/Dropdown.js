import React, { useEffect } from "react";
import FormLabel from "./FormLabel";
import PropTypes from "prop-types";
import classnames from "classnames";
import useUniqueId from "../hooks/useUniqueId";

// Only load USWDS comboBox JS on client-side since it
// references `window`, which isn't available during
// the Node.js-based build process ("server-side")
let comboBox = null;
if (typeof window !== "undefined") {
  comboBox = require("uswds/src/js/components/combo-box");
}

/**
 * A dropdown (`select`) allows users to select one option from a temporary modal menu.
 * Also renders supporting UI elements like a `label`, hint text, and error message.
 *
 * [USWDS Reference ↗](https://designsystem.digital.gov/components/form-controls/#dropdown)
 */
function Dropdown(props) {
  const hasError = !!props.errorMsg;
  const inputId = useUniqueId("Dropdown");
  const { autocomplete } = props;
  useEffect(() => {
    if (autocomplete && comboBox) {
      comboBox.on();

      return () => {
        comboBox.off();
      };
    }
  }, [autocomplete]);

  const fieldClasses = classnames(
    "usa-select maxw-mobile-lg",
    props.selectClassName,
    {
      "usa-input--error": hasError,
    }
  );

  const formGroupClasses = classnames(
    "usa-form-group",
    props.formGroupClassName,
    {
      "usa-form-group--error": hasError,
    }
  );

  const comboBoxClasses = classnames({ "usa-combo-box": autocomplete });

  return (
    <div className={formGroupClasses}>
      <FormLabel
        errorMsg={props.errorMsg}
        hint={props.hint}
        inputId={inputId}
        optionalText={props.optionalText}
        small={props.smallLabel}
        labelClassName={props.labelClassName}
      >
        {props.label}
      </FormLabel>
      <div className={comboBoxClasses} data-default-value={props.value}>
        <select
          className={fieldClasses}
          id={inputId}
          name={props.name}
          onChange={props.onChange}
          value={props.value}
          aria-labelledby={`${inputId}_label ${inputId}_hint`}
        >
          {/* Include a blank initial option which will be chosen if no option has been selected yet */}
          {!props.hideEmptyChoice && (
            <option value="">{props.emptyChoiceLabel}</option>
          )}

          {props.choices.map((choice) => (
            <option key={choice.value} value={choice.value}>
              {choice.label}
            </option>
          ))}
        </select>
      </div>
    </div>
  );
}

Dropdown.propTypes = {
  /**
   * Enable combo-box function, help users select an item from a large list of options.
   * [USWDS Reference ↗](https://designsystem.digital.gov/components/combo-box/)
   */
  autocomplete: PropTypes.bool,
  /**
   * List of choices to be rendered in the dropdown
   */
  choices: PropTypes.arrayOf(
    PropTypes.shape({
      label: PropTypes.oneOfType([PropTypes.number, PropTypes.string])
        .isRequired,
      value: PropTypes.oneOfType([PropTypes.number, PropTypes.string])
        .isRequired,
    })
  ).isRequired,
  /**
   * Localized label for the initially selected option when no value is set
   */
  emptyChoiceLabel: PropTypes.string,
  /**
   * Localized error message. Setting this enables the error state styling.
   */
  errorMsg: PropTypes.node,
  /**
   * Localized hint text
   */
  hint: PropTypes.node,
  /**
   * Additional classes to include on the HTML select
   */
  selectClassName: PropTypes.string,
  /**
   * Override the label's default text-bold class
   */
  labelClassName: PropTypes.string,
  /**
   * Additional classes to include on the containing form group element
   */
  formGroupClassName: PropTypes.string,
  /**
   * Localized label
   */
  label: PropTypes.node.isRequired,
  /**
   * HTML input `name` attribute
   */
  name: PropTypes.string.isRequired,
  /**
   * Localized text indicating this field is optional
   */
  optionalText: PropTypes.node,
  /**
   * HTML input `onChange` attribute
   */
  onChange: PropTypes.func,
  /**
   * Enable the smaller label variant
   */
  smallLabel: PropTypes.bool,
  /**
   * Flag to hide empty choice as first option
   */
  hideEmptyChoice: PropTypes.bool,
  /** The `value` of the selected choice */
  value: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
};

export default Dropdown;
