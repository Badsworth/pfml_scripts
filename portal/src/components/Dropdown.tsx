import React, { useEffect } from "react";
import FormLabel from "./FormLabel";
import classnames from "classnames";
import isBlank from "../utils/isBlank";
import useUniqueId from "../hooks/useUniqueId";

// Only load USWDS comboBox JS on client-side since it
// references `window`, which isn't available during
// the Node.js-based build process ("server-side")
let comboBox: {
  on: () => void;
  off: () => void;
} | null = null;
if (typeof window !== "undefined") {
  comboBox = require("uswds/src/js/components/combo-box");
}

interface DropdownProps {
  /**
   * Enable combo-box function, help users select an item from a large list of options.
   * [USWDS Reference ↗](https://designsystem.digital.gov/components/combo-box/)
   */
  autocomplete?: boolean;
  /**
   * List of choices to be rendered in the dropdown
   */
  choices: Array<{
    label: number | string;
    value: number | string;
  }>;
  /**
   * Localized label for the initially selected option when no value is set
   */
  emptyChoiceLabel?: string;
  /**
   * Localized error message. Setting this enables the error state styling.
   */
  errorMsg?: React.ReactNode;
  /**
   * Localized hint text
   */
  hint?: React.ReactNode;
  /**
   * Additional classes to include on the HTML select
   */
  selectClassName?: string;
  /**
   * Override the label's default text-bold class
   */
  labelClassName?: string;
  /**
   * Additional classes to include on the containing form group element
   */
  formGroupClassName?: string;
  /**
   * Localized label
   */
  label: React.ReactNode;
  /**
   * HTML input `name` attribute
   */
  name: string;
  /**
   * Localized text indicating this field is optional
   */
  optionalText?: React.ReactNode;
  /**
   * HTML input `onChange` attribute
   */
  onChange?: React.ChangeEventHandler<HTMLSelectElement>;
  /**
   * Enable the smaller label variant
   */
  smallLabel?: boolean;
  /**
   * Flag to hide empty choice as first option
   */
  hideEmptyChoice?: boolean;
  /** The `value` of the selected choice */
  value?: number | string;
}

/**
 * A dropdown (`select`) allows users to select one option from a temporary modal menu.
 * Also renders supporting UI elements like a `label`, hint text, and error message.
 *
 * [USWDS Reference ↗](https://designsystem.digital.gov/components/form-controls/#dropdown)
 */
function Dropdown(props: DropdownProps) {
  const hasError = !isBlank(props.errorMsg);
  const inputId = useUniqueId("Dropdown");
  const { autocomplete } = props;

  // We set a unique key for the combobox when the autocomplete value changes, to workaround an issue
  // where the USWDS component wouldn't display the new data-default-value value when it changed.
  // A key forces a remount, which forces the USWDS JS to properly update the displayed value.
  // Learn more about React keys: https://reactjs.org/docs/reconciliation.html#keys
  const key = autocomplete ? props.value : undefined;

  useEffect(() => {
    if (autocomplete && comboBox) {
      comboBox.on();

      return () => {
        if (comboBox) {
          comboBox.off();
        }
      };
    }
    // see reasoning for `key` above. We include props.value here
    // so that the USWDS JS reinitializes anytime the selected value changes.
  }, [autocomplete, props.value]);

  const fieldClasses = classnames(
    "usa-select maxw-mobile-lg",
    props.selectClassName,
    {
      "usa-input--error": hasError,
    }
  );

  const formGroupClasses = classnames(
    "usa-form-group",
    props.formGroupClassName,
    {
      "usa-form-group--error": hasError,
    }
  );

  const comboBoxClasses = classnames({ "usa-combo-box": autocomplete });

  return (
    <div className={formGroupClasses} key={key}>
      <FormLabel
        errorMsg={props.errorMsg}
        hint={props.hint}
        inputId={inputId}
        optionalText={props.optionalText}
        small={props.smallLabel}
        labelClassName={props.labelClassName}
      >
        {props.label}
      </FormLabel>
      <div className={comboBoxClasses} data-default-value={props.value}>
        <select
          className={fieldClasses}
          id={inputId}
          name={props.name}
          onChange={props.onChange}
          value={props.value}
          aria-labelledby={`${inputId}_label ${inputId}_hint ${inputId}_error`}
        >
          {/* Include a blank initial option which will be chosen if no option has been selected yet */}
          {!props.hideEmptyChoice && (
            <option value="">{props.emptyChoiceLabel}</option>
          )}

          {props.choices.map((choice) => (
            <option key={choice.value} value={choice.value}>
              {choice.label}
            </option>
          ))}
        </select>
      </div>
    </div>
  );
}

export default Dropdown;
