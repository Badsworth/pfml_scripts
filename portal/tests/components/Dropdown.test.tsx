import { render, screen } from "@testing-library/react";
import Dropdown from "../../src/components/core/Dropdown";
import React from "react";
import userEvent from "@testing-library/user-event";

const renderComponent = (customProps = {}) => {
  const props = {
    choices: [
      {
        label: "Apple",
        value: "a",
      },
      {
        label: "Banana",
        value: "b",
      },
    ],
    emptyChoiceLabel: "Select an answer",
    label: "Field Label",
    name: "field-name",
    onChange: jest.fn(),
    ...customProps,
  };
  return render(<Dropdown {...props} />);
};

describe("Dropdown", () => {
  it("renders select field with list of options", () => {
    renderComponent();
    expect(screen.getByRole("combobox")).toBeInTheDocument();
    expect(screen.getByRole("combobox")).toMatchSnapshot();
    const [placeholder, optionOne, optionTwo] = screen.getAllByRole("option");
    expect(placeholder).toHaveTextContent("Select an answer");
    expect(optionOne).toHaveValue("a");
    expect(optionTwo).toHaveValue("b");
    expect(screen.getByRole("combobox")).toHaveAccessibleName("Field Label");
  });

  it("enables users to select an option from the dropdown", () => {
    const onChange = jest.fn();
    renderComponent({ onChange });
    userEvent.selectOptions(screen.getByRole("combobox"), ["a"]);
    expect(onChange).toHaveBeenCalled();
  });

  it("can display optionalText in the form label", () => {
    renderComponent({ optionalText: "(optional)" });
    expect(
      screen.getByRole("combobox", { name: "Field Label (optional)" })
    ).toBeInTheDocument();
  });

  it("can display a hint in the form label", () => {
    renderComponent({ hint: "this is a pro tip" });

    expect(
      screen.getByRole("combobox", { name: /this is a pro tip/i })
    ).toBeInTheDocument();
  });

  it("can render the label small", () => {
    renderComponent({ smallLabel: true });
    const label = screen.getByText(/Field Label/);
    expect(label).toHaveClass("usa-label text-bold font-heading-xs measure-5");
  });

  it("can handle errors", () => {
    renderComponent({ errorMsg: "Oh no" });
    const select = screen.getByRole("combobox");
    expect(select).toHaveClass("usa-input--error");
    expect(screen.getByText(/Oh no/)).toBeInTheDocument();
  });

  it("when autocomplete is true, user can click input to see dropdown values", () => {
    renderComponent({ autocomplete: true });
    const input = screen.getByRole("combobox");
    expect(screen.queryByRole("option")).not.toBeInTheDocument();
    userEvent.click(input);
    expect(screen.getAllByRole("option")).toHaveLength(2);
  });

  it("when autocomplete is true, users have option to close out via x in ui", () => {
    renderComponent({ autocomplete: true });
    const [clearButton, toggleDropDownbtn] = screen.getAllByRole("button");
    expect(clearButton).toBeInTheDocument();
    expect(toggleDropDownbtn).toBeInTheDocument();
  });
});
