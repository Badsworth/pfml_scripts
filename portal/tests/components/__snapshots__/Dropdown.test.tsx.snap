// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown renders select field with list of options 1`] = `
<select
  aria-labelledby="Dropdown1_label Dropdown1_hint Dropdown1_error"
  class="usa-select maxw-mobile-lg"
  id="Dropdown1"
  name="field-name"
>
  <option
    value=""
  >
    Select an answer
  </option>
  <option
    value="a"
  >
    Apple
  </option>
  <option
    value="b"
  >
    Banana
  </option>
</select>
`;
